File: algorithms/0004/main.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
  return getMedian(combineNums(nums1, nums2))
}

function combineNums(nums1, nums2) {
  const len1 = nums1.length
  const len2 = nums2.length
  const result = []
  let i = 0
  let j = 0
  for (let k = 0; k < len1 + len2; k++) {
    if (j === len2 || (i < len1 && j < len2 && nums1[i] <= nums2[j])) {
      result.push(nums1[i])
      i += 1
    } else if (i === len1 || (i < len1 && j < len2 && nums1[i] > nums2[j])) {
      result.push(nums2[j])
      j += 1
    }
  }
  return result
}
function getMedian(nums) {
  const len = nums.length
  if (len) {
    if (len % 2) {
      return nums[Math.floor(len / 2)]
    } else {
      return (nums[len / 2 - 1] + nums[len / 2]) / 2
    }
  } else {
    return null
  }
}

module.exports = findMedianSortedArrays

File: algorithms/0004/main.go
package p0004

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	return getMedian(combineNums(nums1, nums2))
}

func combineNums(nums1, nums2 []int) []int {
	len1, i, len2, j := len(nums1), 0, len(nums2), 0
	result := make([]int, len1+len2)
	for k := 0; k < len1+len2; k++ {
		if j == len2 || (i < len1 && j < len2 && nums1[i] <= nums2[j]) {
			result[k] = nums1[i]
			i++
		} else if i == len1 || (i < len1 && j < len2 && nums1[i] > nums2[j]) {
			result[k] = nums2[j]
			j++
		}
	}
	return result
}

func getMedian(nums []int) float64 {
	if l := len(nums); l != 0 {
		if l%2 != 0 {
			return float64(nums[l/2])
		}
		return float64(nums[l/2-1]+nums[l/2]) / 2.0
	}
	panic("the length of slice is 0, no result")
}
